snuba/datasets/query_plan_delegator.py

```diff
@@ -0,0 +1,159 @@
+            ),
+        )
+
+
+class ThreadedQueryPlanExecutionStrategy(QueryPlanExecutionStrategy):
+    def __init__(
+        self,
+        request: Request,
+        query_plan_builders: Mapping[BuilderId, ClickhouseQueryPlanBuilder],
+        selector_func: Callable[[Query], List[BuilderId]],
+        callback_func: Optional[CallbackFunc] = None,
+    ) -> None:
+        self.__request = request
+        self.__query_plan_builders = query_plan_builders
+        self.__selector_func = selector_func
+        self.__callback_func = callback_func
+
+    @with_span()
+    def execute(
+        self, query: Query, request_settings: RequestSettings, runner: QueryRunner
+    ) -> QueryResult:
+        def execute_query(
+            query_plan_builder: ClickhouseQueryPlanBuilder,
+        ) -> QueryResult:
+            query_plan = query_plan_builder.build_plan(self.__request)
+
+            for clickhouse_processor in query_plan.plan_processors:
+                with sentry_sdk.start_span(
+                    description=type(clickhouse_processor).__name__, op="processor"
+                ):
+                    clickhouse_processor.process_query(
+                        query_plan.query, request_settings
+                    )
+
+            return query_plan.execution_strategy.execute(
+                query_plan.query, request_settings, runner
+            )
+
+        def execute_queries() -> Iterator[Tuple[BuilderId, QueryResult]]:
+            builder_ids = self.__selector_func(query)
+            primary_builder_id = builder_ids.pop(0)
+
+            if len(builder_ids) == 0:
+                yield primary_builder_id, execute_query(
+                    self.__query_plan_builders[primary_builder_id]
+                )
+            else:
+                executor = ThreadPoolExecutor()
+
+                with executor:
+                    futures = [
+                        (
+                            builder_id,
+                            executor.submit(
+                                execute_query,
+                                query_plan_builder=self.__query_plan_builders[
+                                    builder_id
+                                ],
+                            ),
+                        )
+                        for builder_id in builder_ids
+                    ]
+
+                    yield primary_builder_id, execute_query(
+                        self.__query_plan_builders[primary_builder_id]
+                    )
+
+                    yield from [
+                        (builder_id, future.result())
+                        for (builder_id, future) in futures
+                    ]
+
+        generator = execute_queries()
+
+        query_results: QueryResults = []
+
+        try:
+            builder_id, query_result = next(generator)
+            query_results.append((builder_id, query_result))
+            return query_result
+        finally:
```

Comment:
How does this work with query errors? What if all the queries fail, will one of the errors get surfaced correctly?

Reply:
@lynnagara this is a good point. Errors when running the query need to be addressed since now we let many errors bubble up and deal with the error in the view.

Reply:
Yeah, this is a good point - only errors in the "primary" query should be raised and be handled the normal way. Otherwise we should just report them to Sentry. I'll tackle this when porting this PR to the query pipeline class.